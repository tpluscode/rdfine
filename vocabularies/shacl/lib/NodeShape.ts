import RdfResourceImpl, { Constructor, namespace, RdfResource } from '@tpluscode/rdfine';
import * as $rdf from '@rdf-esm/data-model';
import type * as RDF from 'rdf-js';
import { sh } from './namespace';
import type { Initializer, ResourceNode } from '@tpluscode/rdfine/RdfResource';
import type { Mixin } from '@tpluscode/rdfine/lib/ResourceFactory';
import type * as Sh from '..';
import { ShapeMixin } from './Shape';

export interface NodeShape<D extends RDF.DatasetCore = RDF.DatasetCore> extends Sh.Shape<D>, RdfResource<D> {
}

export function NodeShapeMixin<Base extends Constructor>(Resource: Base): Constructor<NodeShape> & Base {
  @namespace(sh)
  class NodeShapeClass extends ShapeMixin(Resource) implements NodeShape {
  }
  return NodeShapeClass
}

class NodeShapeImpl extends NodeShapeMixin(RdfResourceImpl) {
  constructor(arg: ResourceNode, init?: Initializer<NodeShape>) {
    super(arg, init)
    this.types.add(sh.NodeShape)
  }

  static readonly __mixins: Mixin[] = [NodeShapeMixin, ShapeMixin];
}
NodeShapeMixin.appliesTo = sh.NodeShape
NodeShapeMixin.Class = NodeShapeImpl
