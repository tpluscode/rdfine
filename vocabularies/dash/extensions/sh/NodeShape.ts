import { ExtendingConstructor, Constructor, namespace, property } from '@tpluscode/rdfine';
import type { RdfResourceCore } from '@tpluscode/rdfine/RdfResource';
import * as $rdf from '@rdf-esm/data-model';
import type * as RDF from 'rdf-js';
import { sh } from '@tpluscode/rdf-ns-builders';
import { dash } from '../../lib/namespace';
import type * as Dash from '../..';
import type * as Shacl from '@rdfine/shacl';
import { ShapeMixin as ShaclShapeMixin } from '@rdfine/shacl/lib/Shape';

interface NodeShapeEx<D extends RDF.DatasetCore = RDF.DatasetCore> {
  defaultViewForRole: Shacl.Shape<D> | undefined;
  shapeScript: Array<Dash.ShapeScript<D>>;
}

declare module '@rdfine/shacl/lib/NodeShape' {
  interface NodeShape<D extends RDF.DatasetCore> extends NodeShapeEx<D> {
  }
}

export function NodeShapeMixinEx<Base extends ExtendingConstructor<Shacl.NodeShape, NodeShapeEx>>(Resource: Base): Constructor<NodeShapeEx & RdfResourceCore> & Base {
  @namespace(dash)
  class Impl extends Resource implements NodeShapeEx {
    @property.resource({ as: [ShaclShapeMixin] })
    defaultViewForRole: Shacl.Shape | undefined;
    @property.resource({ values: 'array', implicitTypes: [dash.ShapeScript] })
    shapeScript!: Array<Dash.ShapeScript>;
  }
  return Impl
}
NodeShapeMixinEx.appliesTo = sh.NodeShape
