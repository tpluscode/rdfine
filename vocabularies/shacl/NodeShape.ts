import RdfResourceImpl, { Constructor, namespace, RdfResource } from '@tpluscode/rdfine';
import * as $rdf from '@rdf-esm/data-model';
import type * as RDF from 'rdf-js';
import { sh } from './lib/namespace';
import type { Initializer, ResourceNode } from '@tpluscode/rdfine/RdfResource';
import type { Mixin } from '@tpluscode/rdfine/lib/ResourceFactory';
import type * as Sh from '.';
import { ShapeMixin } from './Shape';

export interface NodeShape extends Sh.Shape, RdfResource {
}

export function NodeShapeMixin<Base extends Constructor>(Resource: Base) {
  @namespace(sh)
  class NodeShapeClass extends ShapeMixin(Resource) implements NodeShape {
  }
  return NodeShapeClass
}

class NodeShapeImpl extends NodeShapeMixin(RdfResourceImpl) {
  constructor(arg: ResourceNode, init?: Initializer<NodeShape>) {
    super(arg, init)
    this.types.add(sh.NodeShape)
  }

  static readonly __mixins: Mixin[] = [NodeShapeMixin, ShapeMixin];
}
NodeShapeMixin.appliesTo = sh.NodeShape
NodeShapeMixin.Class = NodeShapeImpl
