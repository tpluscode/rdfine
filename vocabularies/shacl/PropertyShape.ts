import RdfResourceImpl, { Constructor, namespace, RdfResource, property } from '@tpluscode/rdfine';
import type * as RDF from 'rdf-js';
import { sh } from './lib/namespace';
import type { Initializer, ResourceNode } from '@tpluscode/rdfine/RdfResource';
import type { Mixin } from '@tpluscode/rdfine/lib/ResourceFactory';
import type * as Sh from '.';
import type * as Rdfs from '@rdfine/rdfs';
import { ResourceMixin as RdfsResourceMixin } from '@rdfine/rdfs/Resource';
import { ShapeMixin } from './Shape';

export interface PropertyShape extends Sh.Shape, RdfResource {
  defaultValue: RDF.Term;
  description: RDF.Term;
  group: Sh.PropertyGroup;
  name: RDF.Term;
  path: Rdfs.Resource;
}

export function PropertyShapeMixin<Base extends Constructor>(Resource: Base) {
  @namespace(sh)
  class PropertyShapeClass extends ShapeMixin(Resource) implements PropertyShape {
    @property()
    defaultValue!: RDF.Term;
    @property()
    description!: RDF.Term;
    @property.resource({ implicitTypes: [sh.PropertyGroup] })
    group!: Sh.PropertyGroup;
    @property()
    name!: RDF.Term;
    @property.resource({ as: [RdfsResourceMixin] })
    path!: Rdfs.Resource;
  }
  return PropertyShapeClass
}

class PropertyShapeImpl extends PropertyShapeMixin(RdfResourceImpl) {
  constructor(arg: ResourceNode, init?: Initializer<PropertyShape>) {
    super(arg, init)
    this.types.add(sh.PropertyShape)
  }

  static readonly __mixins: Mixin[] = [PropertyShapeMixin, ShapeMixin];
}
PropertyShapeMixin.appliesTo = sh.PropertyShape
PropertyShapeMixin.Class = PropertyShapeImpl
