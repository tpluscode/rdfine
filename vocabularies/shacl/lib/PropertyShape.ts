import RdfResourceImpl, { Constructor, namespace, RdfResource, property } from '@tpluscode/rdfine';
import * as $rdf from '@rdf-esm/data-model';
import type * as RDF from 'rdf-js';
import { sh } from './namespace';
import type { Initializer, ResourceNode } from '@tpluscode/rdfine/RdfResource';
import type { Mixin } from '@tpluscode/rdfine/lib/ResourceFactory';
import type * as Sh from '..';
import type * as Rdfs from '@rdfine/rdfs';
import type * as Rdf from '@rdfine/rdf';
import { ShapeMixin } from './Shape';
import { DatatypeMixin as RdfsDatatypeMixin } from '@rdfine/rdfs/lib/Datatype';
import { PropertyMixin as RdfPropertyMixin } from '@rdfine/rdf/lib/Property';
import { ResourceMixin as RdfsResourceMixin } from '@rdfine/rdfs/lib/Resource';

export interface PropertyShape<D extends RDF.DatasetCore = RDF.DatasetCore> extends Sh.Shape<D>, RdfResource<D> {
  datatype: Rdfs.Datatype<D> | undefined;
  defaultValue: RDF.Term | undefined;
  description: string | undefined;
  disjoint: Rdf.Property<D> | undefined;
  '_equals': Rdf.Property<D> | undefined;
  flags: string | undefined;
  group: Sh.PropertyGroup<D> | undefined;
  hasValue: Array<RDF.Term>;
  languageIn: Array<string>;
  lessThan: Rdf.Property<D> | undefined;
  lessThanOrEquals: Rdf.Property<D> | undefined;
  maxCount: number | undefined;
  maxExclusive: number | undefined;
  maxInclusive: number | undefined;
  maxLength: number | undefined;
  minCount: number | undefined;
  minExclusive: number | undefined;
  minInclusive: number | undefined;
  minLength: number | undefined;
  name: string | undefined;
  order: number | undefined;
  path: Rdfs.Resource<D> | Array<Rdfs.Resource<D>> | undefined;
  pattern: string | undefined;
  uniqueLang: boolean | undefined;
}

export function PropertyShapeMixin<Base extends Constructor>(Resource: Base): Constructor<PropertyShape> & Base {
  @namespace(sh)
  class PropertyShapeClass extends ShapeMixin(Resource) implements PropertyShape {
    @property.resource({ as: [RdfsDatatypeMixin] })
    datatype: Rdfs.Datatype | undefined;
    @property()
    defaultValue: RDF.Term | undefined;
    @property.literal()
    description: string | undefined;
    @property.resource({ as: [RdfPropertyMixin] })
    disjoint: Rdf.Property | undefined;
    @property.resource({ as: [RdfPropertyMixin] })
    '_equals': Rdf.Property | undefined;
    @property.literal()
    flags: string | undefined;
    @property.resource({ implicitTypes: [sh.PropertyGroup] })
    group: Sh.PropertyGroup | undefined;
    @property({ values: 'array' })
    hasValue!: Array<RDF.Term>;
    @property.literal({ values: 'list' })
    languageIn!: Array<string>;
    @property.resource({ as: [RdfPropertyMixin] })
    lessThan: Rdf.Property | undefined;
    @property.resource({ as: [RdfPropertyMixin] })
    lessThanOrEquals: Rdf.Property | undefined;
    @property.literal({ type: Number })
    maxCount: number | undefined;
    @property.literal({ type: Number })
    maxExclusive: number | undefined;
    @property.literal({ type: Number })
    maxInclusive: number | undefined;
    @property.literal({ type: Number })
    maxLength: number | undefined;
    @property.literal({ type: Number })
    minCount: number | undefined;
    @property.literal({ type: Number })
    minExclusive: number | undefined;
    @property.literal({ type: Number })
    minInclusive: number | undefined;
    @property.literal({ type: Number })
    minLength: number | undefined;
    @property.literal()
    name: string | undefined;
    @property.literal({ type: Number })
    order: number | undefined;
    @property.resource({ values: ['list', 'single'], as: [RdfsResourceMixin] })
    path: Rdfs.Resource | Array<Rdfs.Resource> | undefined;
    @property.literal()
    pattern: string | undefined;
    @property.literal({ type: Boolean })
    uniqueLang: boolean | undefined;
  }
  return PropertyShapeClass
}

class PropertyShapeImpl extends PropertyShapeMixin(RdfResourceImpl) {
  constructor(arg: ResourceNode, init?: Initializer<PropertyShape>) {
    super(arg, init)
    this.types.add(sh.PropertyShape)
  }

  static readonly __mixins: Mixin[] = [PropertyShapeMixin, ShapeMixin];
}
PropertyShapeMixin.appliesTo = sh.PropertyShape
PropertyShapeMixin.Class = PropertyShapeImpl
