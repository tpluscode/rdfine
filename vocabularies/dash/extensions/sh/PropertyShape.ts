import { ExtendingConstructor, Constructor, namespace, property } from '@tpluscode/rdfine';
import type { RdfResourceCore } from '@tpluscode/rdfine/RdfResource';
import * as $rdf from '@rdf-esm/data-model';
import type * as RDF from 'rdf-js';
import { sh } from '@tpluscode/rdf-ns-builders';
import { dash } from '../../lib/namespace';
import type * as Dash from '../..';
import type * as Shacl from '@rdfine/shacl';
import { NodeShapeMixin as ShaclNodeShapeMixin } from '@rdfine/shacl/lib/NodeShape';

interface PropertyShapeEx<D extends RDF.DatasetCore = RDF.DatasetCore> {
  composite: boolean | undefined;
  editor: Dash.Editor<D> | undefined;
  hidden: boolean | undefined;
  indexed: boolean | undefined;
  mimeTypes: string | undefined;
  readOnly: boolean | undefined;
  reifiableBy: Array<Shacl.NodeShape<D>>;
  viewer: Dash.Viewer<D> | undefined;
}

declare module '@rdfine/shacl/lib/PropertyShape' {
  interface PropertyShape<D extends RDF.DatasetCore> extends PropertyShapeEx<D> {
  }
}

export function PropertyShapeMixinEx<Base extends ExtendingConstructor<Shacl.PropertyShape, PropertyShapeEx>>(Resource: Base): Constructor<PropertyShapeEx & RdfResourceCore> & Base {
  @namespace(dash)
  class Impl extends Resource implements PropertyShapeEx {
    @property.literal({ type: Boolean })
    composite: boolean | undefined;
    @property.resource({ implicitTypes: [dash.Editor] })
    editor: Dash.Editor | undefined;
    @property.literal({ type: Boolean })
    hidden: boolean | undefined;
    @property.literal({ type: Boolean })
    indexed: boolean | undefined;
    @property.literal()
    mimeTypes: string | undefined;
    @property.literal({ type: Boolean })
    readOnly: boolean | undefined;
    @property.resource({ values: 'array', as: [ShaclNodeShapeMixin] })
    reifiableBy!: Array<Shacl.NodeShape>;
    @property.resource({ implicitTypes: [dash.Viewer] })
    viewer: Dash.Viewer | undefined;
  }
  return Impl
}
PropertyShapeMixinEx.appliesTo = sh.PropertyShape
