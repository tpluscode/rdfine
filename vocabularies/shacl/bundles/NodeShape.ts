import type { Mixin } from '@tpluscode/rdfine/lib/ResourceFactory';
import { ShapeMixin } from '../Shape';
import { NodeShapeMixin } from '../NodeShape';
import { NodeKindMixin } from '../NodeKind';

export const NodeShapeBundle = [
  ShapeMixin as Mixin,
  NodeShapeMixin as Mixin,
  NodeKindMixin as Mixin];
