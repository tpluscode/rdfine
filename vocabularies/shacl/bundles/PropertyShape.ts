import type { Mixin } from '@tpluscode/rdfine/lib/ResourceFactory';
import { ShapeMixin } from '../Shape';
import { PropertyGroupMixin } from '../PropertyGroup';
import { PropertyShapeMixin } from '../PropertyShape';

export const PropertyShapeBundle = [
  ShapeMixin as Mixin,
  PropertyGroupMixin as Mixin,
  PropertyShapeMixin as Mixin];
