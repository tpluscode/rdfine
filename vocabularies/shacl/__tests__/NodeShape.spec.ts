import $rdf from 'rdf-ext';
import clownface from 'clownface';
import RDF from '@rdfjs/data-model';
import { fromPointer, NodeShapeMixin } from '../lib/NodeShape.js';
import RdfResourceImpl from '@tpluscode/rdfine';
import { PropertyShapeMixin } from '../lib/PropertyShape.js';
import { rdfs } from '@tpluscode/rdf-ns-builders';
import { expect } from 'chai';
import ns from '@rdfjs/namespace';

const ex = ns('http://example.com/')

RdfResourceImpl.factory.addMixin(NodeShapeMixin)
RdfResourceImpl.factory.addMixin(PropertyShapeMixin)

describe('NodeShape', () => {
  describe('initializer', () => {
    it('sets an array as list deep in the initialized graph', () => {
      // given
      const dataset = $rdf.dataset()
      const graph = clownface({ dataset, term: RDF.namedNode('http://example.com/shape') })

      // when
      fromPointer(graph, {
        property: {
          node: {
            property: [{
              in: [{
                id: ex.foo,
                [rdfs.label.value]: 'Foo',
              }, {
                id: ex.bar,
                [rdfs.label.value]: 'Bar',
              }],
            }],
          },
        },
      })

      // then
      expect(dataset.toCanonical()).toMatchSnapshot()
    })
  })
})
