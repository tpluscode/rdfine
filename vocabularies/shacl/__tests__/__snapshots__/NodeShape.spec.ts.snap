// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NodeShape initializer sets an array as list deep in the initialized graph: toMatchSnapshot 1`] = `
"<http://example.com/bar> <http://www.w3.org/2000/01/rdf-schema#label> \\"Bar\\" .
<http://example.com/foo> <http://www.w3.org/2000/01/rdf-schema#label> \\"Foo\\" .
<http://example.com/shape> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/ns/shacl#NodeShape> .
<http://example.com/shape> <http://www.w3.org/ns/shacl#property> _:c14n4 .
_:c14n0 <http://www.w3.org/1999/02/22-rdf-syntax-ns#first> <http://example.com/bar> .
_:c14n0 <http://www.w3.org/1999/02/22-rdf-syntax-ns#rest> <http://www.w3.org/1999/02/22-rdf-syntax-ns#nil> .
_:c14n1 <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/ns/shacl#NodeShape> .
_:c14n1 <http://www.w3.org/ns/shacl#property> _:c14n2 .
_:c14n2 <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/ns/shacl#PropertyShape> .
_:c14n2 <http://www.w3.org/ns/shacl#in> _:c14n3 .
_:c14n3 <http://www.w3.org/1999/02/22-rdf-syntax-ns#first> <http://example.com/foo> .
_:c14n3 <http://www.w3.org/1999/02/22-rdf-syntax-ns#rest> _:c14n0 .
_:c14n4 <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/ns/shacl#PropertyShape> .
_:c14n4 <http://www.w3.org/ns/shacl#node> _:c14n1 .
"
`;
